packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx

```diff
@@ -29,9 +29,23 @@ export default function useTdpClientCanvas(props: Props) {
     clusterId,
     setTdpConnection,
     setWsConnection,
+    canShareClipboard: canShareClipboardIn,
   } = props;
   const [tdpClient, setTdpClient] = useState<TdpClient | null>(null);
-  const firstImageFragmentRef = useRef(true);
+  const initialTdpConnectionSucceeded = useRef(false);
+  const latestClipboardText = useRef<string>(null);
+
+  // The canShareClipboard prop is used in sendLocalClipboardToRemote to determine
+  // whether or not we actually want to check for and send clipboard data. Because
+  // this information must be fetched from the server (in useDesktopSession), the component
+  // must give us an initial value and update it subsequently. This useRef system below
+  // ensures that canShareClipboard.current gets updated with the latest info from the server,
+  // and that sendLocalClipboardToRemote actually uses that latest value (as opposed to whatever
+  // value the canShareClipboard prop was initialized to, which is what happens in the naive implementation).
+  const canShareClipboard = useRef<boolean>(canShareClipboardIn);
+  if (canShareClipboard.current !== canShareClipboardIn) {
+    canShareClipboard.current = canShareClipboardIn;
+  }
```

Comment:
`useTdpClientCanvas` is called here
https://github.com/gravitational/webapps/blob/e5cf3b92c794968d257f81dfb526d8b10294169f/packages/teleport/src/DesktopSession/useDesktopSession.tsx#L63-L70

where `canShareClipboard` is in part defined by `hasClipboardAccess`
https://github.com/gravitational/webapps/blob/e5cf3b92c794968d257f81dfb526d8b10294169f/packages/teleport/src/DesktopSession/useDesktopSession.tsx#L45 
which is a state variable that's set based on the user's role (computed on the backend and placed in the acl), and not known to us until the component has had time to load and call the api.
https://github.com/gravitational/webapps/blob/e5cf3b92c794968d257f81dfb526d8b10294169f/packages/teleport/src/DesktopSession/useDesktopSession.tsx#L117-L121

Therefore the component is initialized with `hasClipboardAccess` in its default state `false`, which means on the first render `canShareClipboard: hasClipboardAccess && isUsingChrome` is always `false`, and so `canShareClipboardIn` defaults to `false`
https://github.com/gravitational/webapps/blob/e5cf3b92c794968d257f81dfb526d8b10294169f/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx#L32

If I don't use this ref variable here and instead write `sendLocalClipboardToRemote` like this (replacing `canShareClipboard.current` with `canShareClipboardIn`
```typescript
  const sendLocalClipboardToRemote = (cli: TdpClient) => {
    // We must check that the DOM is focused or navigator.clipboard.readText throws an error.
    // We check that initialTdpConnectionSucceeded so that we don't mistakenly send clipboard data
    // to a backend that isn't ready for it yet, which would fail silently.
    if (
      canShareClipboardIn &&
      document.hasFocus() &&
      initialTdpConnectionSucceeded.current
    ) {
      navigator.clipboard.readText().then(text => {
        if (text != latestClipboardText.current) {
          // Wrap in try catch so that lastCopiedClipboardText is only
          // updated if sendClipboardData succeeds.
          // eslint-disable-next-line no-useless-catch
          try {
            cli.sendClipboardData({
              data: text,
            });
            latestClipboardText.current = text;
          } catch (e) {
            throw e;
          }
        }
      });
    }
  };
```

then that `canShareClipboardIn` just gets set to whatever `canShareClipboardIn` was on the first render (when it defaulted to `false`), that clause always evaluates to false and text is never read from the clipboard.

I tried this alternative solution with `useMemo`, however it also fails:
```typescript
  console.log('canShareClipboardIn outside = ', canShareClipboardIn);
  const sendLocalClipboardToRemote = useMemo(() => {
    console.log('canShareClipboardIn in useMemo = ', canShareClipboardIn);
    return (cli: TdpClient) => {
      // We must check that the DOM is focused or navigator.clipboard.readText throws an error.
      // We check that initialTdpConnectionSucceeded so that we don't mistakenly send clipboard data
      // to a backend that isn't ready for it yet, which would fail silently.
      console.log(
        'canShareClipboardIn inside function = ',
        canShareClipboardIn
      );
      console.log('canShareClipboard.current = ', canShareClipboard.current);
      if (
        canShareClipboard.current &&
        document.hasFocus() &&
        initialTdpConnectionSucceeded.current
      ) {
        navigator.clipboard.readText().then(text => {
          if (text != latestClipboardText.current) {
            // Wrap in try catch so that lastCopiedClipboardText is only
            // updated if sendClipboardData succeeds.
            // eslint-disable-next-line no-useless-catch
            try {
              cli.sendClipboardData({
                data: text,
              });
              latestClipboardText.current = text;
            } catch (e) {
              throw e;
            }
          }
        });
      }
    };
  }, [canShareClipboardIn]);
```

the console after I load and induce `sendLocalClipboardToRemote` a few times looks like:
```
canShareClipboardIn outside =  false
app.js:32581 canShareClipboardIn in useMemo =  false
2app.js:32579 canShareClipboardIn outside =  false
app.js:32579 canShareClipboardIn outside =  false
app.js:32579 canShareClipboardIn outside =  true
app.js:32581 canShareClipboardIn in useMemo =  true
app.js:32579 canShareClipboardIn outside =  true
app.js:32579 canShareClipboardIn outside =  true
2app.js:32579 canShareClipboardIn outside =  true
2app.js:32579 canShareClipboardIn outside =  true
app.js:32579 canShareClipboardIn outside =  true
app.js:32579 canShareClipboardIn outside =  true
app.js:32586 canShareClipboardIn inside function =  false
app.js:32587 canShareClipboard.current =  true
app.js:32586 canShareClipboardIn inside function =  false
app.js:32587 canShareClipboard.current =  true
app.js:32586 canShareClipboardIn inside function =  false
app.js:32587 canShareClipboard.current =  true
app.js:32586 canShareClipboardIn inside function =  false
app.js:32587 canShareClipboard.current =  true
app.js:32586 canShareClipboardIn inside function =  false
app.js:32587 canShareClipboard.current =  true
```

So it seems as though JS/React defines that function with the initial value of `canShareClipboardIn`, and can't be prompted to change it by a rerender.

CC @alex-kovoy do you see a better way to do this?

Reply:
oh... i think i know why it's not updating... you send 
```
const onMouseEnter = (cli: TdpClient, e: MouseEvent) => {
    e.preventDefault();
    sendLocalClipboardToRemote(cli);
  };
```
down to `TdpClientCanvas`, which in there you set it in `useEffect` which isn't affected by the state update of `canShareClipboard`:
```
useEffect(() => {
  if (tdpCliOnClipboardData) {
        tdpCli.on(TdpClientEvent.TDP_CLIPBOARD_DATA, (data: ClipboardData) => {
          tdpCliOnClipboardData(data);
        });
      }
}, [tdpCli])
```

So it's either you send `canShareClipboard` all the way to `TdpClientCanvas` and have it like below, or do the ref as you did.
```
useEffect(() => {
  if (tdpCliOnClipboardData) {
        tdpCli.on(TdpClientEvent.TDP_CLIPBOARD_DATA, (data: ClipboardData) => {
          tdpCliOnClipboardData(data);
        });
      }
}, [canShareClipboard])
```

Reply:
Nailed it, that was the reason. I can solve it by wrapping each of those passed-in functions in their own `useEffect` down in `TdpClientCanvas`, e.g. `onMouseEnter` will have its own `useEffect` like
```typescript
useEffect(() => {
  const canvas = canvasRef.current;
  const _onmouseenter = (e: MouseEvent) => {
    onMouseEnter(tdpCli, e);
  };
  if (onMouseEnter) {
    canvas.onmouseenter = _onmouseenter;
  }

  return () => {
    if (onMouseEnter)
      canvas.removeEventListener('onmouseenter', _onmouseenter);
  };
}, [onMouseEnter]);
```

I'm going to try that, since it makes `TdpClientCanvas` respond to changes in its props as a user would naturally expect it should, namely if the prop changes then the behavior that prop codes for should change too.

Reply:
Well spotted @kimlisa 👏👏👏